atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -1451,19 +1492,132 @@ private static SnapshotTransaction unwrapSnapshotTransaction(Transaction transac
+                            metricsManager,
+                            keyValueServiceWrapper.apply(keyValueService, counter),
+                            timelockService,
+                            transactionService,
+                            NoOpCleaner.INSTANCE,
+                            () -> startTimestamp,
+                            getConflictDetectionManager(),
+                            SweepStrategyManagers.createDefault(keyValueService),
+                            startTimestamp,
+                            Optional.empty(),
+                            PreCommitConditions.NO_OP,
+                            AtlasDbConstraintCheckingMode.NO_CONSTRAINT_CHECKING,
+                            null,
+                            TransactionReadSentinelBehavior.THROW_EXCEPTION,
+                            false,
+                            timestampValidationReadCache,
+                            getRangesExecutor,
+                            defaultGetRangesConcurrency,
+                            sweepQueueWriter,
+                            deleteExecutor,
+                            validateLocksOnReads,
+                            () -> TRANSACTION_CONFIG),
+                    counter);
+        }
+
+        @Override
+        protected Transaction createTransaction(
+                long immutableTimestamp,
+                Supplier<Long> startTimestampSupplier,
+                LockToken immutableTsLock,
+                PreCommitCondition preCommitCondition) {
+            AtomicInteger counter = new AtomicInteger();
+            return transactionWrapper.apply(
+                    new SerializableTransaction(
+                            metricsManager,
+                            keyValueService,
+                            timelockService,
+                            transactionService,
+                            cleaner,
+                            startTimestampSupplier,
+                            getConflictDetectionManager(),
+                            sweepStrategyManager,
+                            immutableTimestamp,
+                            Optional.of(immutableTsLock),
+                            preCommitCondition,
+                            constraintModeSupplier.get(),
+                            cleaner.getTransactionReadTimeoutMillis(),
+                            TransactionReadSentinelBehavior.THROW_EXCEPTION,
+                            allowHiddenTableAccess,
+                            timestampValidationReadCache,
+                            getRangesExecutor,
+                            defaultGetRangesConcurrency,
+                            sweepQueueWriter,
+                            deleteExecutor,
+                            validateLocksOnReads,
+                            transactionConfig),
+                    counter);
+        }
+    }
+
+    private static class VerifyingDelegate implements AutoDelegate_KeyValueService {
```

Comment:
```suggestion
    private static class VerifyingDelegate implements ForwardingKeyValueService {
```

otherwise this could break workflows reliant on `getDelegates`